src/main/java/org/opensearch/sdk/BaseExtensionRestHandler.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.sdk;
+
+import static org.opensearch.rest.RestStatus.NOT_FOUND;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.opensearch.extensions.rest.ExtensionRestRequest;
+import org.opensearch.extensions.rest.ExtensionRestResponse;
+import org.opensearch.rest.RestHandler.Route;
+
+/**
+ * Provides convenience methods to reduce boilerplate code in an {@link ExtensionRestHandler} implementation.
+ */
+public abstract class BaseExtensionRestHandler implements ExtensionRestHandler {
+
+    /**
+     * Defines a list of methods which handle each rest {@link Route}.
+     *
+     * @return a list of {@link RouteHandler} with corresponding methods to each route.
+     */
+    protected abstract List<RouteHandler> routeHandlers();
+
+    @Override
+    public List<Route> routes() {
+        return List.copyOf(routeHandlers());
+    }
+
+    @Override
+    public ExtensionRestResponse handleRequest(ExtensionRestRequest request) {
+        Optional<RouteHandler> handler = routeHandlers().stream()
+            .filter(rh -> rh.getMethod().equals(request.method()))
+            .filter(rh -> restPathMatches(request.path(), rh.getPath()))
+            .findFirst();
+        return handler.isPresent() ? handler.get().handleRequest(request) : unhandledRequest(request);
+    }
+
+    /**
+     * Determines if the request's path is a match for the configured handler path.
+     *
+     * @param requestPath The path from the {@link ExtensionRestRequest}
+     * @param handlerPath The path from the {@link RouteHandler}
+     * @return true if the request path matches the route
+     */
+    private boolean restPathMatches(String requestPath, String handlerPath) {
+        // Check exact match
+        if (handlerPath.equals(requestPath)) {
+            return true;
+        }
+        // Split path to evaluate named params
+        String[] handlerSplit = handlerPath.split("/");
+        String[] requestSplit = requestPath.split("/");
+        if (handlerSplit.length != requestSplit.length) {
+            return false;
+        }
+        for (int i = 0; i < handlerSplit.length; i++) {
+            if (!(handlerSplit[i].equals(requestSplit[i]) || (handlerSplit[i].startsWith("{") && handlerSplit[i].endsWith("}")))) {
```

Comment:
What's the significance of the 2nd case here?

Reply:
the two cases are:
 - handler string exactly matches request string
 - handler string is a named param, such as `{detector_id}`, which matches any string.

Reply:
Will named param has curly braces around? I thought it's already removed when passing the request

Reply:
The param is exactly what is in the `routes()` (or `routeHandlers()`) method, with the curly braces.   So for a route `GET` `/detectors/{detectorId}` the `handlerSplit` will be   `[ "detectors", "{detectorId}" ]`.  But the user submitting the request will have sent `GET` `/extensions/_anomaly_detection/detectors/ha7jrT78wK` so the `requestSplit` wlil be `[ "detectors", "ha7jrT78wK" ]`.

So we are splitting both strings by `/`, verifying they both have length 2, then checking each element. Index 0 is an exact string match, will match the first clause.  Index 1 is not an exact string match, but the handler starts and ends with curly braces so it passes, and the handler for this combination is filtered in.  

See the test here:
https://github.com/opensearch-project/opensearch-sdk-java/blob/69c8b392c0289285213067ed0d50954f67402a0e/src/test/java/org/opensearch/sdk/sample/helloworld/rest/TestRestHelloAction.java#L68

https://github.com/opensearch-project/opensearch-sdk-java/blob/69c8b392c0289285213067ed0d50954f67402a0e/src/test/java/org/opensearch/sdk/sample/helloworld/rest/TestRestHelloAction.java#L103


When the request is passed, it will have something

Reply:
Thanks for the detailed information!